File: android/src/main/java/app/capgo/widgetkit/CapgoNativeWidgetBridge.java
package app.capgo.widgetkit;

import android.content.Context;
import android.content.SharedPreferences;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

public class CapgoNativeWidgetBridge {

    public static final String DIRECTION_APP_TO_WIDGET = "appToWidget";
    public static final String DIRECTION_WIDGET_TO_APP = "widgetToApp";

    private static final String SESSION_IDS_KEY = "nativeSessionIds";
    private static final String MESSAGE_IDS_KEY = "nativeMessageIds";

    private final SharedPreferences preferences;

    public CapgoNativeWidgetBridge(final Context context) {
        preferences = context.getApplicationContext().getSharedPreferences(CapgoWidgetKitConstants.PREFS_NAME, Context.MODE_PRIVATE);
    }

    public JSONObject startSession(final String requestedWidgetId, final String kind, final JSONObject state, final JSONObject metadata)
        throws JSONException {
        final long nowMs = System.currentTimeMillis();
        final String widgetId =
            requestedWidgetId != null && !requestedWidgetId.isEmpty() ? requestedWidgetId : TemplateRuntime.createIdentifier("widget");
        final JSONObject session = new JSONObject()
            .put("widgetId", widgetId)
            .put("kind", kind == null ? JSONObject.NULL : kind)
            .put("state", TemplateJsonUtils.deepCopyValue(state != null ? state : new JSONObject()))
            .put("metadata", TemplateJsonUtils.deepCopyValue(metadata != null ? metadata : new JSONObject()))
            .put("status", "active")
            .put("createdAt", nowMs)
            .put("updatedAt", nowMs)
            .put("revision", 1);
        saveSession(session);
        return session;
    }

    public JSONObject updateSession(final String widgetId, final JSONObject state, final JSONObject metadata, final boolean merge)
        throws JSONException {
        final JSONObject session = loadSession(widgetId);
        if (session == null) {
            return null;
        }

        if (state != null) {
            session.put("state", merge ? mergeObjects(session.optJSONObject("state"), state) : TemplateJsonUtils.deepCopyObject(state));
        }
        if (metadata != null) {
            session.put(
                "metadata",
                merge ? mergeObjects(session.optJSONObject("metadata"), metadata) : TemplateJsonUtils.deepCopyObject(metadata)
            );
        }
        session.put("status", "active");
        session.put("updatedAt", System.currentTimeMillis());
        session.put("revision", session.optInt("revision") + 1);
        saveSession(session);
        return session;
    }

    public void stopSession(final String widgetId, final JSONObject state) throws JSONException {
        final JSONObject session = loadSession(widgetId);
        if (session == null) {
            return;
        }
        if (state != null) {
            session.put("state", TemplateJsonUtils.deepCopyObject(state));
        }
        session.put("status", "stopped");
        session.put("updatedAt", System.currentTimeMillis());
        session.put("revision", session.optInt("revision") + 1);
        saveSession(session);
    }

    public JSONObject loadSession(final String widgetId) throws JSONException {
        final String raw = preferences.getString(sessionKey(widgetId), null);
        return raw == null ? null : new JSONObject(raw);
    }

    public JSONArray listSessions() throws JSONException {
        final List<JSONObject> sessions = new ArrayList<>();
        for (String widgetId : listIds(SESSION_IDS_KEY)) {
            final JSONObject session = loadSession(widgetId);
            if (session != null) {
                sessions.add(session);
            }
        }
        sessions.sort((left, right) -> Long.compare(right.optLong("updatedAt"), left.optLong("updatedAt")));
        final JSONArray result = new JSONArray();
        for (JSONObject session : sessions) {
            result.put(session);
        }
        return result;
    }

    public JSONObject sendMessage(
        final String widgetId,
        final String direction,
        final String name,
        final JSONObject payload,
        final boolean expectsResponse
    ) throws JSONException {
        final long nowMs = System.currentTimeMillis();
        final JSONObject message = new JSONObject()
            .put("messageId", TemplateRuntime.createIdentifier("message"))
            .put("widgetId", widgetId)
            .put("direction", normalizeDirection(direction))
            .put("name", name)
            .put("payload", payload == null ? JSONObject.NULL : TemplateJsonUtils.deepCopyObject(payload))
            .put("expectsResponse", expectsResponse)
            .put("status", "pending")
            .put("createdAt", nowMs)
            .put("acknowledgedAt", JSONObject.NULL)
            .put("completedAt", JSONObject.NULL)
            .put("response", JSONObject.NULL)
            .put("error", JSONObject.NULL);
        saveMessage(message);
        return message;
    }

    public JSONArray listMessages(
        final String widgetId,
        final String direction,
        final boolean unacknowledgedOnly,
        final boolean pendingOnly
    ) throws JSONException {
        final List<JSONObject> messages = new ArrayList<>();
        final String normalizedDirection = direction == null ? null : normalizeDirection(direction);
        for (String messageId : listIds(MESSAGE_IDS_KEY)) {
            final JSONObject message = loadMessage(messageId);
            if (message == null) {
                continue;
            }
            if (widgetId != null && !widgetId.equals(message.optString("widgetId"))) {
                continue;
            }
            if (normalizedDirection != null && !normalizedDirection.equals(message.optString("direction"))) {
                continue;
            }
            if (unacknowledgedOnly && !message.isNull("acknowledgedAt")) {
                continue;
            }
            if (pendingOnly && !"pending".equals(message.optString("status"))) {
                continue;
            }
            messages.add(message);
        }
        messages.sort((left, right) -> Long.compare(right.optLong("createdAt"), left.optLong("createdAt")));
        final JSONArray result = new JSONArray();
        for (JSONObject message : messages) {
            result.put(message);
        }
        return result;
    }

    public void acknowledgeMessages(final JSONArray messageIds, final String widgetId, final String direction) throws JSONException {
        final Set<String> targetIds = toStringSet(messageIds);
        final String normalizedDirection = direction == null ? null : normalizeDirection(direction);
        final long nowMs = System.currentTimeMillis();
        for (String currentMessageId : listIds(MESSAGE_IDS_KEY)) {
            final JSONObject message = loadMessage(currentMessageId);
            if (message == null) {
                continue;
            }
            final boolean matchesMessageId = !targetIds.isEmpty() && targetIds.contains(message.optString("messageId"));
            final boolean matchesWidget = widgetId != null && widgetId.equals(message.optString("widgetId"));
            final boolean matchesDirection = normalizedDirection == null || normalizedDirection.equals(message.optString("direction"));
            if ((matchesMessageId || matchesWidget) && matchesDirection) {
                message.put("acknowledgedAt", nowMs);
                saveMessage(message);
            }
        }
    }

    public JSONObject completeMessage(final String messageId, final JSONObject response, final String error) throws JSONException {
        final JSONObject message = loadMessage(messageId);
        if (message == null) {
            return null;
        }
        message.put("status", error == null ? "completed" : "failed");
        message.put("completedAt", System.currentTimeMillis());
        message.put("response", response == null ? JSONObject.NULL : TemplateJsonUtils.deepCopyObject(response));
        message.put("error", error == null ? JSONObject.NULL : error);
        saveMessage(message);
        return message;
    }

    private void saveSession(final JSONObject session) throws JSONException {
        final String widgetId = session.getString("widgetId");
        preferences.edit().putString(sessionKey(widgetId), session.toString()).apply();
        final List<String> ids = listIds(SESSION_IDS_KEY);
        if (!ids.contains(widgetId)) {
            ids.add(widgetId);
            saveIds(SESSION_IDS_KEY, ids);
        }
    }

    private JSONObject loadMessage(final String messageId) throws JSONException {
        final String raw = preferences.getString(messageKey(messageId), null);
        return raw == null ? null : new JSONObject(raw);
    }

    private void saveMessage(final JSONObject message) throws JSONException {
        final String messageId = message.getString("messageId");
        preferences.edit().putString(messageKey(messageId), message.toString()).apply();
        final List<String> ids = listIds(MESSAGE_IDS_KEY);
        if (!ids.contains(messageId)) {
            ids.add(messageId);
            saveIds(MESSAGE_IDS_KEY, ids);
        }
    }

    private JSONObject mergeObjects(final JSONObject base, final JSONObject patch) throws JSONException {
        final JSONObject merged = base == null ? new JSONObject() : TemplateJsonUtils.deepCopyObject(base);
        final IteratorWrapper keys = new IteratorWrapper(patch.keys());
        while (keys.hasNext()) {
            final String key = keys.next();
            final Object current = merged.opt(key);
            final Object value = patch.opt(key);
            if (current instanceof JSONObject currentObject && value instanceof JSONObject patchObject) {
                merged.put(key, mergeObjects(currentObject, patchObject));
            } else {
                merged.put(key, TemplateJsonUtils.deepCopyValue(value));
            }
        }
        return merged;
    }

    private Set<String> toStringSet(final JSONArray values) {
        final Set<String> result = new HashSet<>();
        if (values == null) {
            return result;
        }
        for (int index = 0; index < values.length(); index += 1) {
            final String value = values.optString(index, null);
            if (value != null && !value.isEmpty()) {
                result.add(value);
            }
        }
        return result;
    }

    private List<String> listIds(final String key) {
        final String raw = preferences.getString(key, null);
        final List<String> ids = new ArrayList<>();
        if (raw == null) {
            return ids;
        }
        try {
            final JSONArray values = new JSONArray(raw);
            for (int index = 0; index < values.length(); index += 1) {
                final String value = values.optString(index, null);
                if (value != null && !value.isEmpty()) {
                    ids.add(value);
                }
            }
        } catch (JSONException ignored) {
            return new ArrayList<>();
        }
        return ids;
    }

    private void saveIds(final String key, final List<String> ids) {
        final JSONArray values = new JSONArray();
        for (String id : ids) {
            values.put(id);
        }
        preferences.edit().putString(key, values.toString()).apply();
    }

    private String normalizeDirection(final String direction) {
        return DIRECTION_WIDGET_TO_APP.equals(direction) ? DIRECTION_WIDGET_TO_APP : DIRECTION_APP_TO_WIDGET;
    }

    private String sessionKey(final String widgetId) {
        return "native-session:" + widgetId;
    }

    private String messageKey(final String messageId) {
        return "native-message:" + messageId;
    }

    private static final class IteratorWrapper {

        private final java.util.Iterator<String> iterator;

        IteratorWrapper(final java.util.Iterator<String> iterator) {
            this.iterator = iterator;
        }

        boolean hasNext() {
            return iterator.hasNext();
        }

        String next() {
            return iterator.next();
        }
    }
}

File: ios/Sources/CapgoWidgetKitPlugin/CapgoNativeWidgetBridge.swift
import Foundation

public enum CapgoWidgetMessageDirection: String, Codable, Sendable {
    case appToWidget
    case widgetToApp
}

public enum CapgoWidgetMessageStatus: String, Codable, Sendable {
    case pending
    case completed
    case failed
}

public struct StoredWidgetSessionEnvelope: Codable, Hashable, Sendable {
    public var widgetId: String
    public var kind: String?
    public var stateData: Data
    public var metadataData: Data?
    public var status: String
    public var createdAtMs: Int64
    public var updatedAtMs: Int64
    public var revision: Int

    public init(
        widgetId: String,
        kind: String?,
        stateData: Data,
        metadataData: Data?,
        status: String,
        createdAtMs: Int64,
        updatedAtMs: Int64,
        revision: Int
    ) {
        self.widgetId = widgetId
        self.kind = kind
        self.stateData = stateData
        self.metadataData = metadataData
        self.status = status
        self.createdAtMs = createdAtMs
        self.updatedAtMs = updatedAtMs
        self.revision = revision
    }
}

public struct StoredWidgetBridgeMessage: Codable, Hashable, Sendable {
    public var messageId: String
    public var widgetId: String
    public var direction: CapgoWidgetMessageDirection
    public var name: String
    public var payloadData: Data?
    public var expectsResponse: Bool
    public var status: CapgoWidgetMessageStatus
    public var createdAtMs: Int64
    public var acknowledgedAtMs: Int64?
    public var completedAtMs: Int64?
    public var responseData: Data?
    public var error: String?

    public init(
        messageId: String,
        widgetId: String,
        direction: CapgoWidgetMessageDirection,
        name: String,
        payloadData: Data?,
        expectsResponse: Bool,
        status: CapgoWidgetMessageStatus,
        createdAtMs: Int64,
        acknowledgedAtMs: Int64?,
        completedAtMs: Int64?,
        responseData: Data?,
        error: String?
    ) {
        self.messageId = messageId
        self.widgetId = widgetId
        self.direction = direction
        self.name = name
        self.payloadData = payloadData
        self.expectsResponse = expectsResponse
        self.status = status
        self.createdAtMs = createdAtMs
        self.acknowledgedAtMs = acknowledgedAtMs
        self.completedAtMs = completedAtMs
        self.responseData = responseData
        self.error = error
    }
}

public final class CapgoNativeWidgetBridge {
    private let store: NativeWidgetBridgeStore

    public init(bundle: Bundle = .main) throws {
        store = NativeWidgetBridgeStore(defaults: try CapgoWidgetKitSharedStore.defaults(bundle: bundle))
    }

    public func startSession(
        widgetId requestedWidgetId: String? = nil,
        kind: String? = nil,
        stateObject: [String: Any] = [:],
        metadataObject: [String: Any]? = nil,
        nowMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000)
    ) throws -> StoredWidgetSessionEnvelope {
        let widgetId: String
        if let requestedWidgetId, !requestedWidgetId.isEmpty {
            widgetId = requestedWidgetId
        } else {
            widgetId = TemplateRuntime.createIdentifier(prefix: "widget")
        }
        let session = StoredWidgetSessionEnvelope(
            widgetId: widgetId,
            kind: kind,
            stateData: try TemplateRuntime.jsonData(from: stateObject),
            metadataData: try metadataObject.map(TemplateRuntime.jsonData(from:)),
            status: "active",
            createdAtMs: nowMs,
            updatedAtMs: nowMs,
            revision: 1
        )
        try store.saveSession(session)
        return session
    }

    public func updateSession(
        widgetId: String,
        stateObject: [String: Any]?,
        metadataObject: [String: Any]?,
        merge: Bool = false,
        nowMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000)
    ) throws -> StoredWidgetSessionEnvelope? {
        guard var session = store.loadSession(widgetId: widgetId) else {
            return nil
        }

        if let stateObject {
            let current = try TemplateRuntime.jsonObject(from: session.stateData) as? [String: Any] ?? [:]
            session.stateData = try TemplateRuntime.jsonData(from: merge ? Self.merge(current, with: stateObject) : stateObject)
        }
        if let metadataObject {
            let current = try session.metadataData.map(TemplateRuntime.jsonObject(from:)) as? [String: Any] ?? [:]
            session.metadataData = try TemplateRuntime.jsonData(from: merge ? Self.merge(current, with: metadataObject) : metadataObject)
        }
        session.status = "active"
        session.updatedAtMs = nowMs
        session.revision += 1
        try store.saveSession(session)
        return session
    }

    public func stopSession(widgetId: String, stateObject: [String: Any]? = nil, nowMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000)) throws {
        guard var session = store.loadSession(widgetId: widgetId) else {
            return
        }
        if let stateObject {
            session.stateData = try TemplateRuntime.jsonData(from: stateObject)
        }
        session.status = "stopped"
        session.updatedAtMs = nowMs
        session.revision += 1
        try store.saveSession(session)
    }

    public func loadSession(widgetId: String) -> StoredWidgetSessionEnvelope? {
        store.loadSession(widgetId: widgetId)
    }

    public func listSessions() -> [StoredWidgetSessionEnvelope] {
        store.listSessions()
    }

    public func sendMessage(
        widgetId: String,
        name: String,
        direction: CapgoWidgetMessageDirection = .appToWidget,
        payloadObject: [String: Any]? = nil,
        expectsResponse: Bool = false,
        nowMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000)
    ) throws -> StoredWidgetBridgeMessage {
        let message = StoredWidgetBridgeMessage(
            messageId: TemplateRuntime.createIdentifier(prefix: "message"),
            widgetId: widgetId,
            direction: direction,
            name: name,
            payloadData: try payloadObject.map(TemplateRuntime.jsonData(from:)),
            expectsResponse: expectsResponse,
            status: .pending,
            createdAtMs: nowMs,
            acknowledgedAtMs: nil,
            completedAtMs: nil,
            responseData: nil,
            error: nil
        )
        try store.saveMessage(message)
        return message
    }

    public func listMessages(
        widgetId: String? = nil,
        direction: CapgoWidgetMessageDirection? = nil,
        unacknowledgedOnly: Bool = false,
        pendingOnly: Bool = false
    ) -> [StoredWidgetBridgeMessage] {
        store.listMessages().filter { message in
            if let widgetId, message.widgetId != widgetId { return false }
            if let direction, message.direction != direction { return false }
            if unacknowledgedOnly, message.acknowledgedAtMs != nil { return false }
            if pendingOnly, message.status != .pending { return false }
            return true
        }
    }

    public func acknowledgeMessages(
        messageIds: [String]? = nil,
        widgetId: String? = nil,
        direction: CapgoWidgetMessageDirection? = nil,
        acknowledgedAtMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000)
    ) throws {
        let targetIds = Set(messageIds ?? [])
        for var message in store.listMessages() {
            let matchesMessageId = !targetIds.isEmpty && targetIds.contains(message.messageId)
            let matchesWidget = widgetId != nil && message.widgetId == widgetId
            let matchesDirection = direction == nil || message.direction == direction
            guard (matchesMessageId || matchesWidget) && matchesDirection else {
                continue
            }
            message.acknowledgedAtMs = acknowledgedAtMs
            try store.saveMessage(message)
        }
    }

    public func completeMessage(
        messageId: String,
        responseObject: [String: Any]? = nil,
        error: String? = nil,
        completedAtMs: Int64 = Int64(Date().timeIntervalSince1970 * 1000)
    ) throws -> StoredWidgetBridgeMessage? {
        guard var message = store.loadMessage(messageId: messageId) else {
            return nil
        }
        message.status = error == nil ? .completed : .failed
        message.completedAtMs = completedAtMs
        message.responseData = try responseObject.map(TemplateRuntime.jsonData(from:))
        message.error = error
        try store.saveMessage(message)
        return message
    }

    public static func serializeSession(_ session: StoredWidgetSessionEnvelope) throws -> [String: Any] {
        [
            "widgetId": session.widgetId,
            "kind": session.kind as Any,
            "state": try TemplateRuntime.jsonObject(from: session.stateData),
            "metadata": try session.metadataData.map(TemplateRuntime.jsonObject(from:)) as Any,
            "status": session.status,
            "createdAt": session.createdAtMs,
            "updatedAt": session.updatedAtMs,
            "revision": session.revision
        ]
    }

    public static func serializeMessage(_ message: StoredWidgetBridgeMessage) throws -> [String: Any] {
        [
            "messageId": message.messageId,
            "widgetId": message.widgetId,
            "direction": message.direction.rawValue,
            "name": message.name,
            "payload": try message.payloadData.map(TemplateRuntime.jsonObject(from:)) as Any,
            "expectsResponse": message.expectsResponse,
            "status": message.status.rawValue,
            "createdAt": message.createdAtMs,
            "acknowledgedAt": message.acknowledgedAtMs as Any,
            "completedAt": message.completedAtMs as Any,
            "response": try message.responseData.map(TemplateRuntime.jsonObject(from:)) as Any,
            "error": message.error as Any
        ]
    }

    private static func merge(_ base: [String: Any], with patch: [String: Any]) -> [String: Any] {
        var merged = base
        for (key, value) in patch {
            if let currentObject = merged[key] as? [String: Any], let patchObject = value as? [String: Any] {
                merged[key] = merge(currentObject, with: patchObject)
            } else {
                merged[key] = value
            }
        }
        return merged
    }
}

private final class NativeWidgetBridgeStore {
    private static let sessionIdsKey = "capgo.widgetkit.native.session-ids"
    private static let messageIdsKey = "capgo.widgetkit.native.message-ids"

    private let defaults: UserDefaults
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    init(defaults: UserDefaults) {
        self.defaults = defaults
    }

    func loadSession(widgetId: String) -> StoredWidgetSessionEnvelope? {
        guard let data = defaults.data(forKey: sessionKey(widgetId)) else {
            return nil
        }
        return try? decoder.decode(StoredWidgetSessionEnvelope.self, from: data)
    }

    func saveSession(_ session: StoredWidgetSessionEnvelope) throws {
        defaults.set(try encoder.encode(session), forKey: sessionKey(session.widgetId))
        var ids = sessionIds()
        if !ids.contains(session.widgetId) {
            ids.append(session.widgetId)
            defaults.set(ids, forKey: Self.sessionIdsKey)
        }
        defaults.synchronize()
    }

    func listSessions() -> [StoredWidgetSessionEnvelope] {
        sessionIds().compactMap(loadSession(widgetId:))
            .sorted { $0.updatedAtMs > $1.updatedAtMs }
    }

    func loadMessage(messageId: String) -> StoredWidgetBridgeMessage? {
        guard let data = defaults.data(forKey: messageKey(messageId)) else {
            return nil
        }
        return try? decoder.decode(StoredWidgetBridgeMessage.self, from: data)
    }

    func saveMessage(_ message: StoredWidgetBridgeMessage) throws {
        defaults.set(try encoder.encode(message), forKey: messageKey(message.messageId))
        var ids = messageIds()
        if !ids.contains(message.messageId) {
            ids.append(message.messageId)
            defaults.set(ids, forKey: Self.messageIdsKey)
        }
        defaults.synchronize()
    }

    func listMessages() -> [StoredWidgetBridgeMessage] {
        messageIds().compactMap(loadMessage(messageId:))
            .sorted { $0.createdAtMs > $1.createdAtMs }
    }

    private func sessionIds() -> [String] {
        defaults.stringArray(forKey: Self.sessionIdsKey) ?? []
    }

    private func messageIds() -> [String] {
        defaults.stringArray(forKey: Self.messageIdsKey) ?? []
    }

    private func sessionKey(_ widgetId: String) -> String {
        "capgo.widgetkit.native.session.\(widgetId)"
    }

    private func messageKey(_ messageId: String) -> String {
        "capgo.widgetkit.native.message.\(messageId)"
    }
}
